### ISSUE TITLE
[stable/prometheus-operator] Alert rules are created for components even if scrapes are disabled

### ISSUE THREAD

**Issue by Evesy** [helm/charts#8749]
*Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Feature Request

**Version of Helm and Kubernetes**:
*Helm*: 2.10.0
*Kubernetes*: 1.10.6

**Which chart**:
stable/prometheus

**What happened**:
If scraping of a particular Kubernetes components (e.g. API server) is disabled via Helm values, corresponding alert rules are still configured in Prometheus leading to alerts that are constantly firing.

**What you expected to happen**:
If scraping configurations are disabled then alert rules that would rely on those metrics should also be disabled.

**How to reproduce it** (as minimally and precisely as possible):
- Deploy the prometheus operator chart with the below values set:
  ```
  kubeControllerManager:
    enabled: false
  ```
- The below alert rule is still configured in the  `PrometheusRule` resource and will immediately enter an alerting state:
  ```
    - alert: KubeControllerManagerDown
      annotations:
        message: KubeControllerManager has disappeared from Prometheus target discovery.
        runbook_url: https://github.com/kubernetes-monitoring/kubernetes-mixin/tree/master/runbook.md#alert-name-kubecontrollermanagerdown
      expr: |
        absent(up{job="kube-controller-manager"} == 1)
      for: 15m
      labels:
        severity: critical
  ```

**Comment by vsliouniaev** [comment]
Thanks @Evesy, this is definitely something I was planning on looking at, though I'm not sure how quickly I can get to it. As a workaround, you can copy out whatever rules you want into a separate `prometheusRule` resource and configure `prometheus.prometheusSpec.ruleSelector` to pick it up: https://github.com/helm/charts/blob/master/stable/prometheus-operator/values.yaml#L669

**Comment by Evesy** [comment]
Hey @vsliouniaev -- I'm happy to work on a PR for this if you want to agree on an implementation?

I think a simple conditional on rules that rely on metricset `x` depending on if they're `enabled`/`disabled` is all it would need, thoughts?

**Comment by vsliouniaev** [comment]
This is the way it worked on the coreos charts, I think it's a good approach in the general case, but there may be variations:

We have an etcd and node exporter scrape configuration using `additionalscrapeconfigs` because of how our infrastructure is set up. In this scenario the rules and the exporter aren't necessarily tied to one another. To avoid having to manage the rules using two separate charts we used to just ignore the rules that were deployed with the exporters and roll our own configuration separately. 

In the old chart architecture, it would be weirder to have on off switches for rules without deploying the corresponding exporter/service/servivemonitor. It would also be hard (impossible?) to have a global off switch for rules without using `global`.

TL;DR Sounds good, but I'd like to keep the global off switch of this makes sense. 

@gianrubio, what do you think?

**Comment by paskal** [comment]
One solution would be having something like `defaultRules.kubeScheduler.create` which looks bad if I understood you correctly. Other way could be something like `prometheus.prometheusSpec.defaultRules.kubeScheduler.enable` which seems much better to me. When there will be some decision on this matter, I'll be happy create/test PR (in case @Evesy will prepare it).
By the way, this repo lost all mentions on how to configure prometheus-operator on GKE, was it intentional?

**Comment by vsliouniaev** [comment]
How about just an option to turn off the rules entirely, like it exists now?

**Comment by paskal** [comment]
In general I'd prefer to be able to configure these without copying everything to localhost and losing ability to stay up-to-date with upstream. In reality I've just started configuring these and already see what there are quite a few things to be adjusted in alerting and maybe maintaining the local copy is the only sane way.
I think it's better to ask someone with bigger experience, I have no vote power here but just glad to provide an extra pair of hands for such small adjustments. If you'll see anything prometheus-related where task is clear but you don't want to do it yourself, please cast me and I'll try to help.

**Comment by vsliouniaev** [comment]
In this case, I think it would be better to have an option to disable all rules, would you agree?

**Comment by paskal** [comment]
Yes, but I thought `defaultRules.create` is already doing so?

**Comment by vsliouniaev** [comment]
Ah yes, I'm just requesting that we keep it ;-)

**Comment by paskal** [comment]
If I've got it correctly, one way is to do nothing and other is to create a ton of options to disable particular default alerts so the user will be able to keep up-to-date community-maintained alerting. If you'll want this to go this way, ping me and I'll do it (but then we'll need to know where such options should live).

**Comment by paskal** [comment]
By the way, it seems like having multiply alert files will be more consistent with multiply dashboard files. That way also we could divide them into logical parts and each will be wrapped in it's own if statement.

**Comment by vsliouniaev** [comment]
@paskal Having additional dashboards doesn't automatically cause alerts to trigger, so I think it's ok to leave them in for now. Since dashboards are loaded from files and certain dashboards rely on multiple exporters being available I think it's simpler to leave them enabled, rather than creating additional complexity around their configuration.

If I was to implement this I would break out alerts so that deploying an exporter would be required to enable an alert. I would also have a global on/off switch for the rules as it exists now

**Comment by paskal** [comment]
I'll prepare PR which will separate alerts by files by logical parts (kubelet alerts, etcd alerts, and so on) and enable them with something like `{{- if and .Values.defaultRules.create .Values.kubelet.enabled }}`.

As per discussion in slack, additional variables to turn these on and off make everything too complex, so in scenario when someone need to disable data scraping for some element but have it's alerting she will need to add these rules separately.

**Comment by gianrubio** [comment]
> I'll prepare PR which will separate alerts by files by logical parts (kubelet alerts, etcd alerts, and so on) and enable them with something like `{{- if and .Values.defaultRules.create .Values.kubelet.enabled }}`.
> 
> As per discussion in slack, additional variables to turn these on and off make everything too complex, so in scenario when someone need to disable data scraping for some element but have it's alerting she will need to add these rules separately.

please read https://github.com/helm/charts/pull/8908#issuecomment-434428551

**Comment by paskal** [comment]
Moving question from closed merge request there:
@gianrubio, can I propose separation of [upstream rules](https://github.com/coreos/prometheus-operator/blob/master/contrib/kube-prometheus/manifests/prometheus-rules.yaml) into multiply files, or introduce if statements in existing file without changing anything else?
Also there is an issue with [some non-etcd](https://github.com/helm/charts/blob/master/stable/prometheus-operator/templates/all-prometheus-rules.yaml#L1172) resources located in it's group, I might propose a PR to fix it in upstream, if there is a chance it will be merged.

**Comment by gianrubio** [comment]
@paskal I'd like to have a script similar to [this](https://github.com/coreos/prometheus-operator/blob/master/helm/hack/sync_kube_prometheus.py) that automatically sync the rules from prometheus-operator to helm charts. After we got this automation we can split the rules in multiple files. Do you want to adapt or rewrite this script?

**Comment by paskal** [comment]
@gianrubio, I'm a bit overw**helm**ed. Everything, even etcd [have](https://github.com/etcd-io/etcd/pull/9640) libsonnet file with monitoring rules, what should we do for this chart - collect these files from monitoring-mixin and etcd and build in-place, or use already built version from coreos/prometheus-operator?
@brancz, could you please give your advice as well?

**Comment by brancz** [comment]
The helm charts are entirely community maintained, I sadly can’t help you. We intense use the jsonnet based stack and can answer any question you have about that, if you have anything concrete. :)

**Comment by paskal** [comment]
What is preferable, fetch yaml from jsonnet directly, or take yaml from coreos/prometheus-operator (which in turn fetch it from jsonnet)?

**Comment by brancz** [comment]
I'm not sure I understand, but I recommend you to go through this readme if you are interested in the jsonnet based stack: https://github.com/coreos/prometheus-operator/tree/master/contrib/kube-prometheus#kube-prometheus

**Comment by paskal** [comment]
I've decided to go on with import from coreos/prometheus-operator (instead of importing kubernetes-mixin directly), I'll proceed once I'll get a working computer, two weeks ago my laptop broke and I did not had ability to wrap my head around this issue yet.

**Comment by paskal** [comment]
JIC it's still not resolved, but we got decision how to fix it and will be fixed after my PR merge.
